akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala

```diff
@@ -1803,6 +1803,41 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
       case elem if clazz.isInstance(elem) => supplier.get()
     }: PartialFunction[Throwable, Graph[SourceShape[Out], NotUsed]])
 
+  /**
+   * CompleteOn allows to complete the stream when an upstream error occurs.
+   *
+   * Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements.
+   * This operator can recover the failure signal, but not the skipped elements, which will be dropped.
+   *
+   * '''Emits when''' element is available from the upstream
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or fails with exception is an instance of the provided type
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn(clazz: Class[_ <: Throwable]): javadsl.Flow[In, Out, Mat] = completeOn(ex => clazz.isInstance(ex))
+
+  /**
+   * CompleteOn allows to complete the stream when an upstream error occurs.
+   *
+   * Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements.
+   * This operator can recover the failure signal, but not the skipped elements, which will be dropped.
+   *
+   * '''Emits when''' element is available from the upstream
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or fails with predicate return ture
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn(predicate: java.util.function.Predicate[_ >: Throwable]): javadsl.Flow[In, Out, Mat] =
+    new Flow(delegate.recoverWithRetries(-1, {
+      case ex: Throwable if predicate.test(ex) => scaladsl.Source.empty
+    }))
```

Comment:
Define a name attribute for it and set on the composed stage, also, define in one place (FlowOps) and delegate to that rather than repeat impl in Java DSL etc.